README.md
# stactools-nisar-sim

[![PyPI](https://img.shields.io/pypi/v/stactools-nisar-sim)](https://pypi.org/project/stactools-nisar-sim/)

- Name: nisar-sim
- Package: `stactools.nisar_sim`
- [stactools-nisar-sim on PyPI](https://pypi.org/project/stactools-nisar-sim/)
- Owner: @jjfrench @wildintellect
- [Dataset homepage](https://uavsar.jpl.nasa.gov/science/documents/nisar-sample-products.html)
- STAC extensions used:
  - [proj](https://github.com/stac-extensions/projection/)
  - [sat](https://github.com/stac-extensions/sat)
- [Browse the example in human-readable form](https://radiantearth.github.io/stac-browser/#/external/raw.githubusercontent.com/stactools-packages/nisar-sim/main/examples/collection.json)

stactools package for use with the simulated NISAR products.

## STAC Examples

- [Collection](examples/collection.json)
- [Item](examples/winnip_31604_12061_004_120717_L090_CX_07/winnip_31604_12061_004_120717_L090_CX_07.json)

## Installation

```shell
pip install stactools-nisar-sim
```

## Command-line Usage

Command line functions

```shell
stac nisarsim create-collection <destination/>
```

**Dither**: Indicates whether the data has been dithered or not:

- "X": not dithered,
- "G": dithered with gaps,
- "D": dithered with no gaps.

**Nmode**: NISAR mode, a 3-digit mode number associated a specific center
  frequency, bandwidth, and polarization.

- "129"
- "138"
- "143"

Read more about the _Dither_ and _Nmode_ namings [here](https://uavsar.jpl.nasa.gov/science/documents/nisar-sample-products.html#Product%20Description:~:text=about%20dithered%20products.-,Naming%20Convention,-The%20simulated%20NISAR).

```shell
stac nisarsim create-item <source/> <destination/> --dither <dither> --nmode <nmode>

stac nisarsim create-item
https://downloaduav.jpl.nasa.gov/Release2v/winnip_31604_12061_004_120717_L090_CX_07/
examples/ --dither X --nmode 129
```

Use `stac nisarsim --help` to see all subcommands and options.

## Contributing

We use [pre-commit](https://pre-commit.com/) to check any changes.
To set up your development environment:

```shell
pip install -e .
pip install -r requirements-dev.txt
pre-commit install
```

To check all files:

```shell
pre-commit run --all-files
```

To run the tests:

```shell
./scripts/test
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
This project attempts to match the major and minor versions of
[stactools](https://github.com/stac-utils/stactools) and increments the patch
number as needed.

## [Unreleased]

### Added

- Initial package setup
- Collection creation
- Item creation, for now omitting the SAT extension and avoiding reading the h5py metadata.
- Collection and item examples
- Tests against item and collection generation

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Nothing.

[Unreleased]: <https://github.com/stactools-packages/nisar-sim/tree/main/>


tests/test_stac.py
from stactools.nisar_sim import stac


def test_create_collection() -> None:
    # Write tests for each for the creation of a STAC Collection
    # Create the STAC Collection...
    collection = stac.create_collection()
    collection.set_self_href("")

    # Check that it has some required attributes
    assert collection.id == "nisar-sim"
    # self.assertEqual(collection.other_attr...

    # Validate
    collection.validate()


def test_create_item() -> None:
    # Write tests for each for the creation of STAC Items
    # Create the STAC Item...
    item = stac.create_item(
        "tests/data-files/winnip_31604_12061_004_120717_L090_CX_07",
        dither="X",
        nmode="129",
    )

    # Check that it has some required attributes
    assert item.id == "winnip_31604_12061_004_120717_L090_CX_07"
    # self.assertEqual(item.other_attr...

    # Validate
    item.validate()


tests/test_module.py
import stactools.nisar_sim


def test_version() -> None:
    assert stactools.nisar_sim.__version__ is not None


src/stactools/nisar_sim/stac.py
import logging
import re

from pystac import Collection, Item, Summaries
from pystac.extensions.item_assets import ItemAssetsExtension
from pystac.extensions.projection import ProjectionExtension
from pystac.extensions.sat import SatExtension

from stactools.nisar_sim import constants as c
from stactools.nisar_sim.metadata import (
    AnnotatedMetadata,
    HDF5Metadata,
    Metadata,
    MetadataLinks,
    fill_sat_properties,
)
from stactools.nisar_sim.nisar_assets import get_assets

logger = logging.getLogger(__name__)


def create_collection() -> Collection:
    """Create a STAC Collection

    This function includes logic to extract all relevant metadata from
    an asset describing the STAC collection and/or metadata coded into an
    accompanying constants.py file.

    See `Collection<https://pystac.readthedocs.io/en/latest/api.html#collection>`_.

    Returns:
        Collection: STAC Collection object
    """
    summary_dict = {
        "platform": c.NISAR_SIM_PLATFORMS,
    }

    collection = Collection(
        id="nisar-sim",
        title="Simulated NISAR",
        description=c.NISAR_SIM_DESCRIPTION,
        providers=c.NISAR_SIM_PROVIDERS,
        extent=c.NISAR_SIM_EXTENT,
        summaries=Summaries(summary_dict),
        stac_extensions=[
            ItemAssetsExtension.get_schema_uri(),
            ProjectionExtension.get_schema_uri(),
            SatExtension.get_schema_uri(),
        ],
        keywords=c.NISAR_SIM_KEYWORDS,
    )

    # Links
    collection.add_links(c.NISAR_SIM_LINKS)

    assets = ItemAssetsExtension.ext(collection, add_if_missing=True)
    assets.item_assets = get_assets(collection=True)

    return collection


def create_item(
    product_path: str, dither: str, nmode: str, sat_extension: bool = False
) -> Item:
    """Create a STAC Item

    Args:
        product_href (str): An HREF data directory pointing to a flight
        dither (str): Type of dither ("X","G","D")
            X: No Dither
            G: Dither with gaps
            D: Dithered without gaps
        nmode (str): set of numbers associated with a specific center frequency,
            bandwidth, and polarization ("129","138","143")
        sat_extension (bool): Whether to include the SAT extension.
        Requires that the HDF5 file is present in `product_path`.

    Returns:
        Item: STAC Item object
    """
    metalinks = MetadataLinks(product_path, dither, nmode)
    h5_data = HDF5Metadata(metalinks.h5_href) if sat_extension else None
    ann_data = AnnotatedMetadata(metalinks.ann_a_href, metalinks.ann_b_href)
    metadata = Metadata(product_path, metalinks.id, h5_data, ann_data)

    item = Item(
        id=metalinks.id,
        properties={},
        geometry=metadata.geometry,
        bbox=metadata.bbox,
        datetime=metadata.get_datetime,
        stac_extensions=[],
    )

    proj_attrs = ProjectionExtension.ext(item, add_if_missing=True)
    proj_attrs.epsg = 4326

    get_xtalk = re.search(r"_(\w)\w_", metalinks.id)
    xtalk = get_xtalk[1] if get_xtalk else "C"

    expected_assets = get_assets(dither=dither, xtalk=xtalk)

    # fmt: off
    def asset_name(name: str) -> str:
        versionless = name[:name.rfind("_")] + name[name.find("."):]
        return versionless[len(metadata.base_id):].lstrip("_")
    # fmt: on

    for _file in metadata.inventory:
        if asset_def := expected_assets.get(asset_name(_file)):
            asset = asset_def.create_asset(_file)
            item.add_asset(asset_name(_file), asset)

    # SAT extension
    if sat_extension:
        sat = SatExtension.ext(item, add_if_missing=True)
        if h5_data is not None:
            fill_sat_properties(sat, h5_data.metadata)
        else:
            raise ValueError("HDF5 data not found. Cannot add SAT extension.")

    return item


src/stactools/nisar_sim/metadata.py
import os
import re
from datetime import datetime
from pathlib import Path
from typing import Any, Dict, List, Optional, Tuple, TypeVar

import h5py
import numpy as np
import pystac
from pystac.extensions.sat import SatExtension

T = TypeVar("T", pystac.Item, pystac.Asset)


class H5MetadataException(Exception):
    pass


class AnnotatedMetadataException(Exception):
    pass


class MetadataLinks:
    """Generates the correct paths to metadata files"""

    def __init__(self, href: str, dither: str, nmode: str) -> None:
        self.asset_href = href
        self.dither = dither
        self.nmode = nmode
        self.id = os.path.basename(os.path.normpath(href))

        self.h5_href = self._href_modifier(href, extension=".h5", nmode=nmode)
        self.ann_a_href = self._href_modifier(href, extension=".ann", nmode=f"{nmode}A")
        self.ann_b_href = self._href_modifier(href, extension=".ann", nmode=f"{nmode}B")

    def _replace_dither(self, id: str, dither: str) -> str:
        id_pattern = r"_([CX])([DGX])_"
        return re.sub(id_pattern, f"_\\1{dither}_", id)

    def _href_modifier(
        self,
        href: str,
        extension: Optional[str] = None,
        nmode: Optional[str] = None,
    ) -> str:
        _id = self._replace_dither(self.id, self.dither)
        path = Path(href) / _id

        if extension:
            path = path.with_suffix(extension)
        if nmode is not None:
            path_parts = str(path).split("_")
            path_parts.insert(-1, nmode)
            path = Path("_".join(path_parts))
        return str(path)


class HDF5Metadata:
    """Reads HDF5 file into a usable h5py object"""

    def __init__(self, h5_href: str) -> None:
        self.href = h5_href
        self.metadata = self.from_file(self.href)

    def from_file(self, href: str) -> Any:
        return h5py.File(href, "r")


class AnnotatedMetadata:
    """Reads annotation file into a usable dict"""

    def __init__(self, ann_a_href: str, ann_b_href: str) -> None:
        self.ann_a_href = ann_a_href
        self.ann_b_href = ann_b_href
        self.metadata_a = self.parse_ann(self.ann_a_href)
        self.metadata_b = self.parse_ann(self.ann_b_href)

    def parse_ann(self, ann_href: str) -> Dict[str, Any]:
        """Reads in annotated file location and returns dict of key-value pairs
        Strips out comments, spacing, and units
        Args:
            ann_href (str):

        Example:
            DEM Datum                                         (&)         = WGS-84
            becomes {"DEM_Datum": "WGS-84"}
        """
        with open(ann_href, "r") as f:
            lines = f.readlines()
            data = {}
            for line in lines:
                if "=" in line:
                    key_value = line.split("=", 1)
                    key = (
                        re.sub(r"^[^a-zA-Z]*", "", key_value[0])
                        .split("(")[0]
                        .strip()
                        .replace(" ", "_")
                    )
                    value = key_value[1].split(";")[0].strip()
                    data[key] = value
        return data


class Metadata:
    def __init__(self, href: str, id: str, h5_metadata: Any, ann_metadata: Any) -> None:
        self.href = href
        self.base_id = "_".join(id.split("_")[:-2])
        self.h5_metadata = h5_metadata
        self.ann_metadata = ann_metadata

        def _get_geometries_from_ann_metadata() -> Tuple[List[float], Dict[str, Any]]:
            # using the A file for geometries. A and B are approximately similar.

            def _get_rounded_coord(k: str) -> Any:
                return np.round(float(self.ann_metadata.metadata_a[k]), 3)

            bbox = [
                _get_rounded_coord(k)
                for k in [
                    "Approximate_Lower_Left_Longitude",
                    "Approximate_Lower_Left_Latitude",
                    "Approximate_Upper_Right_Longitude",
                    "Approximate_Upper_Right_Latitude",
                ]
            ]

            polygon_coords = [
                [
                    _get_rounded_coord(k)
                    for k in [
                        "Approximate_Upper_Left_Longitude",
                        "Approximate_Upper_Left_Latitude",
                    ]
                ],
                [
                    _get_rounded_coord(k)
                    for k in [
                        "Approximate_Lower_Left_Longitude",
                        "Approximate_Lower_Left_Latitude",
                    ]
                ],
                [
                    _get_rounded_coord(k)
                    for k in [
                        "Approximate_Lower_Right_Longitude",
                        "Approximate_Lower_Right_Latitude",
                    ]
                ],
                [
                    _get_rounded_coord(k)
                    for k in [
                        "Approximate_Upper_Right_Longitude",
                        "Approximate_Upper_Right_Latitude",
                    ]
                ],
                [
                    _get_rounded_coord(k)
                    for k in [
                        "Approximate_Upper_Left_Longitude",
                        "Approximate_Upper_Left_Latitude",
                    ]
                ],
            ]

            geometry = {"type": "Polygon", "coordinates": [polygon_coords]}
            return (bbox, geometry)

        self.bbox, self.geometry = _get_geometries_from_ann_metadata()

    @property
    def start_datetime(self) -> datetime:
        return datetime.strptime(
            self.ann_metadata.metadata_a["Start_Time_of_Acquisition"],
            "%d-%b-%Y %H:%M:%S %Z",
        )

    @property
    def end_datetime(self) -> datetime:
        return datetime.strptime(
            self.ann_metadata.metadata_a["Stop_Time_of_Acquisition"],
            "%d-%b-%Y %H:%M:%S %Z",
        )

    @property
    def get_datetime(self) -> datetime:
        return datetime.strptime(
            self.ann_metadata.metadata_a["Date_of_Acquisition"], "%d-%b-%Y %H:%M:%S %Z"
        )

    @property
    def inventory(self) -> List[str]:
        files: List[str] = []
        for metadata in [self.ann_metadata.metadata_a, self.ann_metadata.metadata_b]:
            files.extend(v for k, v in metadata.items() if self.base_id in v)
        if not files:
            raise AnnotatedMetadataException(
                "Cannot determine product files using annotated metadata at "
                f"{self.ann_metadata.ann_a_href} or {self.ann_metadata.ann_b_href}"
            )
        files.extend(
            [
                os.path.basename(self.ann_metadata.ann_a_href),
                os.path.basename(self.ann_metadata.ann_b_href),
            ]
        )
        return files


def fill_sat_properties(sat_ext: SatExtension[T], h5_data: Any) -> None:
    """Fills the properties for SAT
    Based on the sat Extension.py
    Args:
        sat_ext (SarExtension): The extension to be populated
        h5_data (Any): h5 file object
    """

    if absolute_orbit := h5_data["science"]["LSAR"]["identification"].get(
        "absoluteOrbitNumber"
    ):
        sat_ext.absolute_orbit = int(absolute_orbit[()])  # stored as numpy.uint32


src/stactools/nisar_sim/constants.py
from datetime import datetime
from typing import Optional

from pystac import Extent, Link, Provider
from pystac import ProviderRole as PR
from pystac import SpatialExtent, TemporalExtent
from pystac.utils import str_to_datetime

NISAR_SIM_COLLECTION_START: Optional[datetime] = str_to_datetime("2008-01-01T00:00:00Z")
NISAR_SIM_TEMPORAL_EXTENT = TemporalExtent([[NISAR_SIM_COLLECTION_START, None]])
NISAR_SIM_SPATIAL_EXTENT = SpatialExtent([-180.0, -90.0, 180.0, 90.0])
NISAR_SIM_EXTENT = Extent(NISAR_SIM_SPATIAL_EXTENT, NISAR_SIM_TEMPORAL_EXTENT)

NISAR_SIM_PROVIDERS = [
    Provider(
        "NASA/JPL-Caltech",
        roles=[PR.PRODUCER, PR.PROCESSOR, PR.LICENSOR],
        url="https://www.jpl.nasa.gov/caltechjpl-privacy-policies-and-important-notices",
    ),
    Provider(
        "NASA/JPL-Caltech",
        roles=[PR.HOST],
        url="https://www.jpl.nasa.gov/caltechjpl-privacy-policies-and-important-notices",
    ),
]
NISAR_SIM_KEYWORDS = ["SAR", "NISAR", "Simulated"]
NISAR_SIM_DESCRIPTION = "Simulated NISAR products are specialized UAVSAR PolSAR products. They include files described in the UAVSAR PolSAR Data Format but with different number of MLC looks and GRD grid; as well as additional NISAR RSLC HDF5 files, radiometric terrain correction layer, and flat earth incidence angle files."  # noqa

NISAR_SIM_LINKS = [
    Link(
        rel="documentation",
        target=(
            "https://uavsar.jpl.nasa.gov/science/documents/nisar-sample-products.html"
        ),
        media_type="application/html",
        title="Simulated NISAR Products documentation",
        extra_fields={
            "description": (
                "Simulated NISAR products are generated from UAVSAR data to emulate NISAR data"
                " characteristics in order to help users test their algorithms and get a sense of"
                " the quality of future NISAR products. Click on the links below to go to the"
                " desired section."
            )
        },
    ),
    Link(
        rel="license",
        target="https://www.jpl.nasa.gov/jpl-image-use-policy",
        title="NASA/JPL-Caltech Image Use Policy",
    ),
]
NISAR_SIM_PLATFORMS = ["NISAR"]
NISAR_SIM_INSTRUMENTS = ["L-band"]
NISAR_SIM_EPSG = 4326

NMODE = ["129", "138", "143"]  # center frequency, bandwidth, and polarization
XTALK = ["X", "C"]


src/stactools/nisar_sim/commands.py
import logging
import os

import click
from click import Command, Group

from stactools.nisar_sim import stac

logger = logging.getLogger(__name__)


def create_nisarsim_command(cli: Group) -> Command:
    """Creates the stactools-nisar-sim command line utility."""

    @cli.group(
        "nisarsim",
        short_help=("Commands for working with stactools-nisar-sim"),
    )
    def nisarsim() -> None:
        pass

    @nisarsim.command(
        "create-collection",
        short_help="Creates a STAC collection",
    )
    @click.argument("destination")
    def create_collection_command(destination: str) -> None:
        """Creates a STAC Collection

        Args:
            destination (str): An HREF for the Collection JSON
        """
        collection = stac.create_collection()

        collection.set_self_href(destination)

        collection.save_object()

        return None

    @nisarsim.command("create-item", short_help="Create a STAC item")
    @click.argument("source")
    @click.argument("destination")
    @click.option(
        "--dither",
        default="X",
        type=click.Choice(["X", "G", "D"], case_sensitive=False),
        help="single character indicating the item's dither type ('X','G','D')",
    )
    @click.option(
        "--nmode",
        default="129",
        type=click.Choice(["129", "138", "143"]),
        help="set of numbers indicating the item's nmode ('129','138','143')",
    )
    @click.option(
        "--sat-extension",
        default=False,
        type=bool,
        help="Whether to include the SAT extension. Requires that the HDF5 file is present in `product_path`.",
    )
    def create_item_command(
        source: str, destination: str, dither: str, nmode: str, sat_extension: bool
    ) -> None:
        """Creates a STAC Item

        Args:
            source (str): HREF of the Asset associated with the Item
            destination (str): An HREF for the STAC Item
            dither (str): single character indicating the item's dither type ("X","G","D")
            nmode (str): set of numbers associated with a specific center frequency,
                bandwidth, and polarization ("129","138","143")
        """
        item = stac.create_item(source, dither, nmode)
        json_file = os.path.basename(source)
        json_path = os.path.join(destination, f"{json_file}.json")
        item.set_self_href(os.path.basename(json_path))

        item.validate()

        item.save_object(dest_href=destination)

        return None

    return nisarsim
